content/oci/deletableoci.go

```diff
@@ -0,0 +1,349 @@
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/errdef"
+	"oras.land/oras-go/v2/internal/container/set"
+	"oras.land/oras-go/v2/internal/descriptor"
+	"oras.land/oras-go/v2/internal/graph"
+	"oras.land/oras-go/v2/internal/resolver"
+)
+
+// DeletableStore implements `oras.Target`, and represents a content store
+// extended with the delete operation.
+// Reference: https://github.com/opencontainers/image-spec/blob/v1.1.0-rc4/image-layout.md
+type DeletableStore struct {
+	// AutoSaveIndex controls if the OCI store will automatically save the index
+	// file on each Tag() call.
+	//   - If AutoSaveIndex is set to true, the OCI store will automatically call
+	//     this method on each Tag() call.
+	//   - If AutoSaveIndex is set to false, it's the caller's responsibility
+	//     to manually call SaveIndex() when needed.
+	//   - Default value: true.
+	AutoSaveIndex bool
+	root          string
+	indexPath     string
+	index         *ocispec.Index
+	lock          sync.RWMutex
+
+	storage     *Storage
+	tagResolver *resolver.Memory
+	graph       *graph.Memory
+}
+
+// NewDeletableStore returns a new DeletableStore with context.Background().
+func NewDeletableStore(root string) (*DeletableStore, error) {
+	return NewDeletableStoreWithContext(context.Background(), root)
+}
+
+// NewDeletableStoreWithContext creates a new DeletableStore.
+func NewDeletableStoreWithContext(ctx context.Context, root string) (*DeletableStore, error) {
+	rootAbs, err := filepath.Abs(root)
+	if err != nil {
+		return nil, fmt.Errorf("failed to resolve absolute path for %s: %w", root, err)
+	}
+	storage, err := NewStorage(rootAbs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create storage: %w", err)
+	}
+	ds := &DeletableStore{
+		AutoSaveIndex: true,
+		root:          rootAbs,
+		indexPath:     filepath.Join(rootAbs, ocispec.ImageIndexFile),
+		storage:       storage,
+		tagResolver:   resolver.NewMemory(),
+		graph:         graph.NewMemory(),
+	}
+	if err := ensureDir(filepath.Join(rootAbs, ocispec.ImageBlobsDir)); err != nil {
+		return nil, err
+	}
+	if err := ds.ensureOCILayoutFile(); err != nil {
+		return nil, fmt.Errorf("invalid OCI Image Layout: %w", err)
+	}
+	if err := ds.loadIndexFile(ctx); err != nil {
+		return nil, fmt.Errorf("invalid OCI Image Index: %w", err)
+	}
+	return ds, nil
+}
+
+// Fetch fetches the content identified by the descriptor.
+func (ds *DeletableStore) Fetch(ctx context.Context, target ocispec.Descriptor) (io.ReadCloser, error) {
+	ds.lock.RLock()
+	defer ds.lock.RUnlock()
+
+	return ds.storage.Fetch(ctx, target)
+}
+
+// Push pushes the content, matching the expected descriptor.
+func (ds *DeletableStore) Push(ctx context.Context, expected ocispec.Descriptor, reader io.Reader) error {
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	if err := ds.storage.Push(ctx, expected, reader); err != nil {
+		return err
+	}
```

Comment:
The current implementation does not allow concurrent pushing as it blocks.

Given that `ds.storage` itself is concurrent-safe, will
```suggestion
	if err := ds.storage.Push(ctx, expected, reader); err != nil {
		return err
	}
	
	ds.lock.Lock()
	defer ds.lock.Unlock()
```
be a better solution?

Reply:
Is the content that `ds.lock` should protect the entire `DeletableStore` or just the metadata (i.e. everything except `ds.storage`)? You may think about it.

Reply:
If there is a concurrent `Delete` operation after the `ds.storage.Push`, we may have problems doing the `Index` and `Tag`. Therefore the lock has to lock the entire `Push`.